x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/TokenService.java

```diff
@@ -235,70 +249,50 @@ public static Boolean isTokenServiceEnabled(Settings settings) {
-                }
-                builder.startObject("access_token")
-                        .field("invalidated", false)
-                        .field("user_token", userToken)
-                        .field("realm", authentication.getAuthenticatedBy().getName())
-                        .endObject();
-                builder.endObject();
-                request = client.prepareIndex(SECURITY_INDEX_NAME, SINGLE_MAPPING_NAME, documentId)
-                                .setOpType(OpType.CREATE)
-                                .setSource(builder)
-                                .setRefreshPolicy(RefreshPolicy.WAIT_UNTIL)
-                                .request();
-            } catch (IOException e) {
-                // unexpected exception
-                listener.onFailure(e);
-                return;
+            final UserToken userToken = new UserToken(userTokenId, version, tokenAuth, getExpirationTime(), metadata);
+            final String plainRefreshToken = includeRefreshToken ? UUIDs.randomBase64UUID() : null;
+            final String versionedRefreshToken = includeRefreshToken ? prependVersionAndEncode(version, plainRefreshToken) : null;
+            final BytesReference tokenDocument;
+            if (version.onOrAfter(VERSION_TOKENS_INDEX_INTRODUCED)) {
+                tokenDocument = createTokenDocument(userToken, plainRefreshToken, originatingClientAuth);
+            } else {
+                tokenDocument = createTokenDocument(userToken, versionedRefreshToken, originatingClientAuth);
             }
-            securityIndex.prepareIndexIfNeededThenExecute(ex -> listener.onFailure(traceLog("prepare security index", documentId, ex)),
-                    () -> executeAsyncWithOrigin(client, SECURITY_ORIGIN, IndexAction.INSTANCE, request, ActionListener
-                            .wrap(indexResponse -> {
+            final String documentId = getTokenDocumentId(userToken);
+            final SecurityIndexManager tokensIndex = getSecurityIndexManagerForVersion(version);
+            final IndexRequest indexTokenRequest = client.prepareIndex(tokensIndex.aliasName(), SINGLE_MAPPING_NAME, documentId)
+                    .setOpType(OpType.CREATE)
+                    .setSource(tokenDocument, XContentType.JSON)
+                    .setRefreshPolicy(RefreshPolicy.WAIT_UNTIL)
+                    .request();
+            tokensIndex.prepareIndexIfNeededThenExecute(ex -> listener.onFailure(traceLog("prepare security tokens index", documentId, ex)),
+                    () -> executeAsyncWithOrigin(client, SECURITY_ORIGIN, IndexAction.INSTANCE, indexTokenRequest,
+                            ActionListener.wrap(indexResponse -> {
                                 if (indexResponse.getResult() == Result.CREATED) {
-                                    listener.onResponse(new Tuple<>(userToken, refreshToken));
+                                    listener.onResponse(new Tuple<>(userToken, versionedRefreshToken));
```

Comment:
I was in the process of annotating the PR and explaining this, but you beat me to it. Thanks for the quick round trip!

It is correct, the versioned refresh token is returned irrespective of the minimum cluster node in the cluster. But it will not fail if it is presented to old nodes.
This is because the refresh token is stored in the versioned format in the `.security` as opposed to unversioned in `security-tokens`. This is so that old nodes can use a refresh token generated by new ones.

**I will detail the strategy in a separate comment.** I think your comment highlighted a bug in another part.

Reply:
> it could get presented to an older node and I think that will fail.

No, because the older node will search the older index and in the old index we store the token as is, ie with the version... I now realize this is odd, although I believe it works. I will change so that new nodes in a mixed cluster behave like old ones and return an unversioned token.

Reply:
This logic has been reworked, the way you'd expected. We always store the unversioned format and prepend the version only on a full upgraded cluster.